src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -51,14 +51,19 @@ public void ScanAndProcessReturnValue (MethodBody methodBody)
 			if (MethodReturnValue != null) {
 				var requiredMemberKinds = _flowAnnotations.GetReturnParameterAnnotation (methodBody.Method);
 				if (requiredMemberKinds != 0) {
-					var reflectionContext = new ReflectionPatternContext (_context, methodBody.Method, methodBody.Method, 0);
+					IMetadataTokenProvider accessedMember = methodBody.Method.MethodReturnType;
+					if (MethodReturnValue is LoadFieldValue field)
+						accessedMember = field.Field;
+					else if (MethodReturnValue is MethodParameterValue param)
+						accessedMember = methodBody.Method.Parameters[param.ParameterIndex - 1];
+					var reflectionContext = new ReflectionPatternContext (_context, methodBody.Method, accessedMember, null);
```

Comment:
```suggestion
					var reflectionContext = new ReflectionPatternContext (_context, methodBody.Method, methodBody.Method.MethodReturnType, null);
```

The problematic spot is not the source of the value but the place where that value is used and needs to meet some requirements - in this case it's being returned from a method which has requirements on the return value.

This is sort of complicated, the experience I would like to see for this is something like:
* The warning gets reported on the return statement
* The warning message contains where the value came from
* The warning contains info about the fact that the value does not fullfill the requirements on the return value

You can try this with nullable - make a method which returns non-nullable reference type and try to return null - and see what the C# analyzer prints out - we should get something very similar.